### PR TITLE
multistage build changes

### LINKED ISSUE #1458: Multistage Docker build
## Describe the bug

Builds are huge and slow

## Expected behavior

Builds should be 🚀 fast and 🐜 small. Using multistage builds could speed things up.

## Steps to Reproduce

`docker build` or downloading the latest image

## Logs

takes `1m 18s` to download image
https://github.com/github/super-linter/runs/2371468980?check_suite_focus=true

## Additional context

This would likely require a fair amount of work on the Dockerfile and the build process. We would need to go through the image layer by layer and pull in only the libs we truly need. This could greatly pull down the size of the image as well but will require deep Docker knowledge and extensive testing...

cc/ @admiralAwkbar @clemenko

### PR DESCRIPTION
**Pull request by zkoppert**
Fixes #1458

## Proposed Changes

1. Build using multistage
2. Validate works
3. Profit?

### CODE REVIEW

**Comment by zkoppert** [comment]
Docker image is not building https://github.com/github/super-linter/pull/1461/checks?check_run_id=2382837532#step:4:2386

**Comment by admiralAwkbar** [comment]
Its building now...
Now time to clean up more... lol

**Comment by admiralAwkbar** [comment]
at this point, we have shaved .5 to 1gb off the size... but there's some more room for cleanup

**Comment by ferrarimarco** [comment]
Nice!

**Comment by admiralAwkbar** [comment]
@clemenko any chance you can look over the `dockerfile` and see if there is anything we can do to pull this in even more?
pathing was a monster... lol

**Review by clemenko — COMMENTED** [review]
Not sure why this is failing. 
`executor failed running [/bin/sh -c ACTIONS_RUNNER_DEBUG=true WRITE_LINTER_VERSIONS_FILE=true /action/lib/linter.sh     && /action/lib/functions/validateDocker.sh]: exit code: 1`

**Comment by clemenko** [comment]
I was able to get it to 1.71gb. roughly 6 from alpine. 900 in /usr/bin and 800 in /usr/lib.
Screen Shot 2021-04-22 at 11 05 01 AM

**Comment by admiralAwkbar** [comment]
@clemenko 
can you post your `dokcerfile`? Thats crazy you got it that small

As you can see from mine, it installs in the following locations:
```
COPY --from=base_image /usr/bin/ /usr/bin/
COPY --from=base_image /usr/local/bin/ /usr/local/bin/
COPY --from=base_image /usr/local/lib/ /usr/local/lib/
COPY --from=base_image /usr/local/share/ /usr/local/share/
COPY --from=base_image /usr/lib /usr/lib/
COPY --from=base_image /usr/share/ /usr/share/
COPY --from=base_image /usr/include/ /usr/include/
COPY --from=base_image /lib/ /lib/
COPY --from=base_image /bin/ /bin/
COPY --from=base_image /opt/microsoft/ /opt/microsoft/
COPY --from=base_image /node_modules/ /node_modules/
COPY --from=base_image /tmp/.rustup/ /tmp/.rustup/
COPY --from=base_image /etc/R/ /etc/R/
```

File: Dockerfile [hunk 1]
```diff
@@ -363,15 +385,11 @@ COPY lib /action/lib
...
-RUN ACTIONS_RUNNER_DEBUG=true WRITE_LINTER_VERSIONS_FILE=true /action/lib/linter.sh
-
-##################################4
-# Run validations of built image #
-##################################
-RUN /action/lib/functions/validateDocker.sh
+################################################
+# Run to build version file and validate image #
+################################################
+RUN ACTIONS_RUNNER_DEBUG=true WRITE_LINTER_VERSIONS_FILE=true /action/lib/linter.sh \
```
[the remark this answers is not in the document]
Comment (admiralAwkbar) [reply Dockerfile:395]: does it show the error below?? that should test that all binaries are available,

**Comment by admiralAwkbar** [comment]
@zkoppert @ferrarimarco this is building but its the tap that fails as we have seen for all builds

this is a great start and we can work to move more layers together and hopefully slim some more
but I think the horrible work has been done ;)

**Review by jwiebalk — APPROVED** [review]

**Pull request merged**